src/3d/qgs3dmodelcache.cpp

```diff
@@ -0,0 +1,157 @@
+/***************************************************************************
+                         qgs3dmodelcache.cpp
+                         -----------------
+    begin                : July 2020
+    copyright            : (C) 2020 by Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dmodelcache.h"
+
+#include "qgis.h"
+#include "qgslogger.h"
+
+#include <QFile>
+#include <QBuffer>
+#include <QTemporaryDir>
+
+///@cond PRIVATE
+
+Qgs3DModelCacheEntry::Qgs3DModelCacheEntry( const QString &path )
+  : QgsAbstractContentCacheEntry( path )
+{
+}
+
+bool Qgs3DModelCacheEntry::isEqual( const QgsAbstractContentCacheEntry *other ) const
+{
+  const Qgs3DModelCacheEntry *other3dModel = dynamic_cast< const Qgs3DModelCacheEntry * >( other );
+  // cheapest checks first!
+  if ( !other3dModel || other3dModel->temporaryFile != temporaryFile )
+    return false;
+
+  return true;
+}
+
+int Qgs3DModelCacheEntry::dataSize() const
+{
+  return QFile( temporaryFile ).size();
+}
+
+void Qgs3DModelCacheEntry::dump() const
+{
+  QgsDebugMsgLevel( QStringLiteral( "path: %1" ).arg( path ), 3 );
+}
+
+///@endcond
+
+// canonical 3D model cache instance
+Qgs3DModelCache *Qgs3DModelCache::s3dModelCache = nullptr;
+
+Qgs3DModelCache::Qgs3DModelCache( QObject *parent )
+  : QgsAbstractContentCache< Qgs3DModelCacheEntry >( parent, QObject::tr( "3D Model" ) )
+{
+  temporaryDir.reset( new QTemporaryDir() );
+
+  connect( this, &QgsAbstractContentCacheBase::remoteContentFetched, this, &Qgs3DModelCache::remote3DModelFetched );
+}
+
+Qt3DRender::QMesh *Qgs3DModelCache::pathAsMesh( const QString &path, bool blocking )
+{
+  Qt3DRender::QMesh *mesh = new Qt3DRender::QMesh;
+
+  const QString file = path.trimmed();
+  if ( file.isEmpty() )
+    return mesh;
+
+  QMutexLocker locker( &mMutex );
```

Comment:
Is it necessary to use mutex here? All creation of QEntity objects is done in the main thread (only loading of actual data to provide geometries/textures is done in worker threads)

Reply:
@wonder-sk , is there no foreseeable scenario where that'll happen off the main thread? I thought it wouldn't hurt to have it there for such a day. If it's not going to happen, we can take it off.

Reply:
I don't see that coming anytime soon - I don't see a reason why we would be going that way. As QEntity objects are derived from QObject, it is in my opinion also cleaner/simpler design to make them always in main thread rather than creating them in worker threads and then moving them to the main thread where they belong (because the scene root entity lives in main thread too). Another reason is that creation of entities should be always fairly quick as all loading should be done in workers run before QEntity gets created.